libs/s25main/mapGenerator/Resources.cpp

```diff
@@ -177,39 +178,73 @@ namespace rttr { namespace mapGenerator {
 
+        // Helper to pick an index [0,4) into the mRIs array that identifies a randomly selected resource we should
+        // place. Note that we do not perform budget adjustments and checks here.
+        auto resourcePicker = [&]() {
+            int ratio = 0;
+            int randomNumber = rnd.RandomValue(1, std::max(1, total));
+            for(int i = 0; i < 4; ++i)
+            {
+                ratio += mRIs[i].ratio;
+                if(randomNumber < ratio)
+                    return i;
+            }
+            return -1;
+        };
+
         RTTR_FOREACH_PT(MapPoint, map.size)
         {
             if(textures.All(pt, IsMinableMountain))
             {
-                int randomNumber = rnd.RandomValue(1, total);
-                int ratio = settings.ratioGold;
-
-                if(randomNumber < ratio)
-                {
-                    resources[pt] = libsiedler2::R_Gold + rnd.RandomValue(0, 8);
+                // Pick a random resource, -1 indicates none.
+                int randomMRIindex = resourcePicker();
+                if(randomMRIindex < 0)
                     continue;
-                }
 
-                ratio += settings.ratioCoal;
-                if(randomNumber < ratio)
-                {
-                    resources[pt] = libsiedler2::R_Coal + rnd.RandomValue(0, 8);
+                auto &mRI = mRIs[randomMRIindex];
+
+                // Adjust and check the budget, if we are over we will not place the resource because we have already in
+                // the past.
+                ++mRI.budget;
+                if(mRI.budget <= 0)
                     continue;
-                }
 
-                ratio += settings.ratioIron;
-                if(randomNumber < ratio)
-                {
-                    resources[pt] = libsiedler2::R_Iron + rnd.RandomValue(0, 8);
+                // Budget is positive so we want to place resources, however, we avoid overwriting existing clusters as
+                // this results in sprinkles. If we have a resource on this field we keep the budget and place a cluster
+                // later.
+                if(resources[pt])
                     continue;
-                }
 
-                ratio += settings.ratioGranite;
-                if(randomNumber < ratio)
+                // Cluster the resource around the map point.
+                for(int xd = -4; xd <= 4; ++xd)
                 {
-                    resources[pt] = libsiedler2::R_Granite + rnd.RandomValue(0, 8);
+                    for(int yd = -4; yd <= 4; ++yd)
+                    {
+                        MapPoint pt_d = MakeMapPoint(Position(pt.x + xd, pt.y + yd), map.size);
+                        // Only place it on mines that have no resource yet, adjust the budget for each placed
+                        // resource.
+                        if(textures.All(pt_d, IsMinableMountain) && !resources[pt_d])
```

Comment:
Switching the conditions should make this quite a bit faster:
```suggestion
                        if(!resources[pt_d] && textures.All(pt_d, IsMinableMountain))
```